### PR TITLE
Update danbooru21_extract.py

### PR DESCRIPTION
**Pull request by chavinlo**
Added:

- Option to disable tags to human-readable text conversion

- Option to specify JSON file, img folder path and saving path from CLI arguments

### CODE REVIEW

File: scripts/danbooru21_extract.py [hunk 1]
```diff
@@ -55,7 +106,8 @@ def convert_rating_to_humanrating(input):
 for json_str in json_list:
     current_line_count = current_line_count + 1
     ##415627 last line of 00.json, ignore
-    print("Current Line:" + str(current_line_count) + '/415627 | Current saved files count: ' + str(current_saved_file_count) )
+    ##TODO: Add a line counter to print progress accurately
+    print("Current Line:" + str(current_line_count) + '/415000 (aprox) | Current saved files count: ' + str(current_saved_file_count) )
```
Comment (JohannesGaessler) [inline scripts/danbooru21_extract.py:110]: Using tqdm seems like less effort to be honest.

**Comment by chavinlo** [comment]
Sorry so much about all this mess, I really forgot how branches work

**Review by JohannesGaessler — COMMENTED** [review]
Please consider rearranging tags based on category.
The Danbooru2021 metadata divides tags into several categories like "tag_string_general" or "tag_string_character".
As it is the tags from all of these categories are simply put in alphabetical order.
However, since the beginnings of prompts seem to be given more weight than the latter parts I would suspect that putting characters and artists at the front would allow for more faithful recreations.

I think a blacklist of bad tags would also be helpful to filter out tags like "md5_mismatch" that are probably just adding noise.

In any case, please document the way in which captions are generated more precisely.
The HuggingFace page simply says "Captions are Danbooru-style captions" which is not very helpful.
If possible, please provide a link to the exact code and revision that was used.
This will make my life as a downstream developer considerably easier if I want to for example implement code that generates a prompt from a Danbooru link.

**Comment by chavinlo** [comment]
Hello @JohannesGaessler , I'm actually the guy you talked to on matrix a few weeks ago. Happy to see you again!!!!
Thank you for the feedback, I will implement the features you mentioned.

Characters = tag_string_character

Artists = tag_string_artist

Show/Anime/Series = tag_string_copyright

Should we put the copyright tag too? And in which order do you suggest? I think Characters, Copyright and Artist would be a good order. Also is there a list of blacklisted tags? Maybe we could check the [danbooru meta tags page](https://danbooru.donmai.us/tags?commit=Search&search%5Bcategory%5D=5&search%5Bhide_empty%5D=yes&search%5Border%5D=count).

And about the hugging-face models, it was trained by haru who **didnt used this script at all**. This script was made right after he started training and was originally meant for another guide (jack 000 guide aka the first rentry training guide). You could ask @harubaru either here or through the [TouhouAI discord](https://discord.gg/7w3rBsSkrj).

I will try to use tqdm, thank you for the suggestion.

**Comment by chavinlo** [comment]
Also, I think he also just used the general tag string

**Comment by JohannesGaessler** [comment]
I'm not sure what would be the best order for tag categories; I think this is a subjective decision that depends on what the finetune is supposed to do (if my intuition is correct in the first place).
My personal priority would be tag_string_character > tag_string_artist > tag_string_general > tag_string_copyright > tag_string_meta.
On second thought I'm not sure if putting artists at the front would be the best move though; the general wisdom with SD prompts is to but them at the back (I didn't rigorously test this but some people have reported that the beginning and end of prompts is given more weights than the middle).
I think it may be worthwhile to look at the order of equivalent terms in the LAION dataset; if the same order is used I would assume the results would be best.

**Comment by harubaru** [comment]
Hey @JohannesGaessler, so far this is how the captions are generated:
`[Copyright Tags] + [Character Tags] + [General Tags] + [Artist Tags]`
Meta tags are not included and having the tags alphabetically sorted is definitely an issue. I was thinking about shuffling all of the tags in a caption but I'm not sure how effective this would be. Would love to hear more thoughts and ideas about this for the next finetune run!

**Comment by harubaru** [comment]
Here is an example caption used during training:
`fire_emblem fire_emblem_fates setsuna_(fire_emblem) 1girl ? artist_name blue_eyes blue_theme blush hair_over_one_eye hairband ilya_kuvshinov_(style) looking_at_viewer monochrome portrait short_hair simple_background sketch solo watermark web_address white_background koyorin`

**Comment by JohannesGaessler** [comment]
Thanks for the clarification.

>I was thinking about shuffling all of the tags in a caption but I'm not sure how effective this would be.

Randomizing tag order might reduce overfitting (since it's effectively data augmentation) though I don't know how much of a problem that is for finetuning.
I think randomization will also increase the compute needed for training though.
In any case, whether tags are shuffled should be clearly communicated to users since it affects how prompts should be constructed.
As I said, please publish the relevant code along with the weights.

I don't know whether there is also a CLIP token limit during training, may lead to some tags being cut off.
If tags are sorted alphabetically the finetune may be comparatively bad at tags that start with X, Y, or Z since those are the most likely to be cut off.

Intuitively I would assume that using punctuation to separate tags will lead to better results compared to just using the format in the Danbooru2021 metadata.
Stable Diffusion does take punctuation into account and commas and periods seems to be interpreted as a separation between concepts.
For example, "Snapchat selfie" is interpreted as a selfie as seen on Snapchat (only one or two people in frame) while "Snapchat, selfie" is interpreted as a mixture of Snapchat and selfie (mostly selfie).

Stable diffusion does differentiate between underscores and whitespaces but they seem to be interpreted very similarly so I don't know whether replacing them would lead to better results.

I am assuming based on your posts that the rating of the images is not taken into account at all for caption generation (though I think to remember someone saying that Waifu Diffusion was only trained on the SFW pre-processed subset of Danbooru2021 anyways).

**Comment by harubaru** [comment]
> As I said, please publish the relevant code along with the weights.

Will do! Sorry for any confusion about the data.

> I don't know whether there is also a CLIP token limit during training, may lead to some tags being cut off.

https://github.com/harubaru/waifu-diffusion/blob/41c864dce2d5398c24f3cb9906faaa505614867d/ldm/modules/encoders/modules.py#L80-L82

There is a 77 token limit. If this tag shuffling idea is implemented, it'll be important to include copyright, artist, and character tags are included in the truncated captions.

**Comment by harubaru** [comment]
@JohannesGaessler Added the scripts used for downloading the data here: 3a84dc1973090291e13e84b3e218ac0d13a7c671

**Comment by JohannesGaessler** [comment]
Thanks.
I'll take a look at it.

In other news: someone on /g/ reported that adding "bad_anatomy" to the negative prompt improves results.
Perhaps this tag should be blacklisted?
